File: desktop-mend/src/test/java/commands/DecryptTest.java
package commands;

import co.samco.mend4.core.AppProperties;
import co.samco.mend4.core.exception.CorruptSettingsException;
import co.samco.mend4.core.exception.MalformedLogFileException;
import co.samco.mend4.desktop.commands.Decrypt;
import co.samco.mend4.desktop.exception.MendLockedException;
import co.samco.mend4.desktop.exception.SettingRequiredException;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.mockito.ArgumentCaptor;

import javax.crypto.BadPaddingException;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.NoSuchPaddingException;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.security.InvalidAlgorithmParameterException;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.security.spec.InvalidKeySpecException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import static org.mockito.Matchers.any;
import static org.mockito.Matchers.eq;
import static org.mockito.Mockito.*;

public class DecryptTest extends TestBase {
    private Decrypt decrypt;

    @Before
    public void setup() {
        super.setup();
        decrypt = new Decrypt(log, strings, settingsHelper, cryptoHelper, osDao, fileResolveHelper);
    }

    @Test
    public void testCommandWithSettingsDependencies() throws IOException, SettingRequiredException {
        super.testCommandWithSettingsDependencies(decrypt);
    }

    @Test
    public void decryptLog() throws IOException, CorruptSettingsException, MalformedLogFileException, InvalidKeyException,
            NoSuchAlgorithmException, IllegalBlockSizeException, BadPaddingException, NoSuchPaddingException,
            InvalidAlgorithmParameterException, InvalidKeySpecException, MendLockedException {
        String logFileName = "logFile." + AppProperties.LOG_FILE_EXTENSION;
        decryptLog(logFileName, false);
    }

    @Test
    public void decryptLogWithSilentFlag() throws IOException, CorruptSettingsException, MalformedLogFileException,
            InvalidKeyException, NoSuchAlgorithmException, IllegalBlockSizeException, BadPaddingException,
            NoSuchPaddingException, InvalidAlgorithmParameterException, InvalidKeySpecException, MendLockedException {
        String logFileName = "logFile." + AppProperties.LOG_FILE_EXTENSION;
        decryptLog(logFileName, true);
    }

    @Test
    public void decryptEnc() throws IOException, CorruptSettingsException, MalformedLogFileException,
            InvalidAlgorithmParameterException, NoSuchAlgorithmException, NoSuchPaddingException, BadPaddingException,
            IllegalBlockSizeException, InvalidKeyException, InvalidKeySpecException, MendLockedException {
        String encFileName = "sam." + AppProperties.ENC_FILE_EXTENSION;
        decryptEnc(encFileName, Arrays.asList(encFileName), false);
    }

    @Test
    public void decryptEncSilent() throws IOException, CorruptSettingsException, MalformedLogFileException,
            InvalidAlgorithmParameterException, NoSuchAlgorithmException, NoSuchPaddingException, BadPaddingException,
            IllegalBlockSizeException, InvalidKeyException, InvalidKeySpecException, MendLockedException {
        String encFileName = "sam." + AppProperties.ENC_FILE_EXTENSION;
        decryptEnc(encFileName, Arrays.asList(encFileName, Decrypt.SILENT_FLAG), true);
    }

    @Test
    public void logFileDoesntExist() throws FileNotFoundException {
        String fileName = "unkown.extension";
        String exceptionText = "hi";
        doThrow(new FileNotFoundException(exceptionText)).when(fileResolveHelper)
                .assertFileExistsAndHasExtension(anyString(), anyString(), any(File.class));
        decrypt.execute(Arrays.asList(fileName));
        ArgumentCaptor<String> errCapture = ArgumentCaptor.forClass(String.class);
        verify(err).println(errCapture.capture());
        Assert.assertEquals(exceptionText, errCapture.getValue());
    }

    @Test
    public void encFileDoesntExist() throws IOException, CorruptSettingsException {
        String fileName = "unkown.extension";
        when(fileResolveHelper.resolveAsEncFilePath(fileName)).thenReturn(new File(fileName));
        when(fileResolveHelper.fileExistsAndHasExtension(anyString(), any(File.class))).thenReturn(false);
        decrypt.execute(Arrays.asList(fileName));
        verify(fileResolveHelper).resolveAsLogFilePath(eq(fileName));
    }

    @Test
    public void noArgs() {
        decrypt.execute(Collections.emptyList());
        ArgumentCaptor<String> errCapture = ArgumentCaptor.forClass(String.class);
        verify(err, times(2)).println(errCapture.capture());
        Assert.assertEquals(strings.get("Decrypt.noFile"), errCapture.getAllValues().get(0));
        Assert.assertEquals(strings.getf("Decrypt.usage", Decrypt.COMMAND_NAME, Decrypt.SILENT_FLAG), errCapture.getAllValues().get(1));
    }

    private void decryptEnc(String encFileName, List<String> args, boolean silentFlag) throws IOException, CorruptSettingsException,
            MalformedLogFileException, InvalidKeyException, NoSuchAlgorithmException, NoSuchPaddingException, BadPaddingException,
            IllegalBlockSizeException, InvalidAlgorithmParameterException, InvalidKeySpecException, MendLockedException {
        when(fileResolveHelper.resolveAsEncFilePath(encFileName)).thenReturn(new File(encFileName));
        when(fileResolveHelper.fileExistsAndHasExtension(eq(AppProperties.ENC_FILE_EXTENSION), any(File.class))).thenReturn(true);
        decrypt.execute(args);
        ArgumentCaptor<File> fileCaptor = ArgumentCaptor.forClass(File.class);
        verify(fileResolveHelper).fileExistsAndHasExtension(eq(AppProperties.ENC_FILE_EXTENSION), any(File.class));
        verify(cryptoHelper).decryptFile(fileCaptor.capture(), eq(silentFlag));
        Assert.assertEquals(encFileName, fileCaptor.getValue().getName());
    }

    private void decryptLog(String logFileName, boolean silentFlag) throws IOException, CorruptSettingsException,
            MalformedLogFileException, InvalidAlgorithmParameterException, NoSuchAlgorithmException, IllegalBlockSizeException,
            BadPaddingException, NoSuchPaddingException, InvalidKeyException, InvalidKeySpecException, MendLockedException {
        when(fileResolveHelper.resolveAsLogFilePath(anyString())).thenReturn(new File(logFileName));
        List<String> args = new ArrayList<>();
        args.add(logFileName);
        if (silentFlag) {
            args.add(Decrypt.SILENT_FLAG);
        }
        decrypt.execute(args);
        ArgumentCaptor<File> fileCaptor = ArgumentCaptor.forClass(File.class);
        verify(cryptoHelper).decryptLog(fileCaptor.capture());
        Assert.assertEquals(logFileName, fileCaptor.getValue().getName());
    }
}

File: desktop-mend/src/test/java/commands/DecryptTest.kt
package commands

import co.samco.mend4.core.AppProperties
import co.samco.mend4.desktop.commands.Decrypt
import co.samco.mend4.desktop.dao.SettingsDao
import com.nhaarman.mockitokotlin2.whenever
import org.junit.Assert.assertEquals
import org.junit.Before
import org.junit.Test
import org.mockito.ArgumentMatchers.anyString
import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.argumentCaptor
import com.nhaarman.mockitokotlin2.doAnswer
import com.nhaarman.mockitokotlin2.eq
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.times
import com.nhaarman.mockitokotlin2.verify
import java.io.File
import java.io.FileNotFoundException

class DecryptTest : TestBase() {
    private lateinit var decrypt: Decrypt

    @Before
    override fun setup() {
        super.setup()
        decrypt = Decrypt(
            log = log,
            strings = strings,
            settingsHelper = settingsHelper,
            cryptoHelper = cryptoHelper,
            fileResolveHelper = fileResolveHelper,
            settings = settings,
        )
    }

    @Test
    fun testCommandWithSettingsDependencies() {
        super.testCommandWithNoSettingsDependencies(decrypt)
    }

    @Test
    fun decryptLog() {
        val logFileName = "logFile." + AppProperties.LOG_FILE_EXTENSION
        whenever(keyHelper.privateKey).thenReturn(mock())
        whenever(
            fileResolveHelper.fileExistsAndHasExtension(
                eq(AppProperties.ENC_FILE_EXTENSION),
                any()
            )
        )
            .thenReturn(false)
        whenever(fileResolveHelper.resolveAsLogFilePath(anyString()))
            .thenReturn(File(logFileName))

        decrypt.execute(listOf(logFileName))

        val fileCaptor = argumentCaptor<File>()
        verify(cryptoHelper).decryptLog(fileCaptor.capture())
        assertEquals(logFileName, fileCaptor.firstValue.name)
    }

    @Test
    fun decryptEnc() {
        val encFileName = "sam." + AppProperties.ENC_FILE_EXTENSION
        decryptEnc(encFileName, listOf(encFileName), false)
    }

    @Test
    fun decryptEncSilent() {
        val encFileName = "sam." + AppProperties.ENC_FILE_EXTENSION
        decryptEnc(encFileName, listOf(Decrypt.SILENT_FLAG, encFileName), true)
    }

    @Test
    fun logFileDoesntExist() {
        val fileName = "unknown.extension"
        val exception = "exception test"

        doAnswer { throw FileNotFoundException(exception) }
            .whenever(fileResolveHelper)
            .assertFileExistsAndHasExtension(
                anyString(),
                anyString(),
                //Bit of a hack using an argument captor otherwise mockito doesn't match the call.
                argumentCaptor<File>().capture()
            )

        decrypt.execute(listOf(fileName))

        verify(err).println(eq(exception))
    }

    @Test
    fun encFileDoesntExist() {
        val fileName = "unkown.extension"
        whenever(fileResolveHelper.resolveAsEncFilePath(fileName)).thenReturn(File(fileName))
        whenever(fileResolveHelper.fileExistsAndHasExtension(anyString(), any()))
            .thenReturn(false)
        decrypt.execute(listOf(fileName))
        verify(fileResolveHelper).resolveAsLogFilePath(eq(fileName))
    }

    @Test
    fun noArgs() {
        decrypt.execute(emptyList())
        val errCapture = argumentCaptor<String>()
        verify(err, times(2)).println(errCapture.capture())
        assertEquals(strings["Decrypt.noFile"], errCapture.allValues[0])
        assertEquals(
            strings.getf(
                "Decrypt.usage",
                Decrypt.COMMAND_NAME,
                Decrypt.SILENT_FLAG
            ), errCapture.allValues[1]
        )
    }

    private fun decryptEnc(encFileName: String, args: List<String>, silentFlag: Boolean) {
        whenever(fileResolveHelper.resolveAsEncFilePath(encFileName))
            .thenReturn(File(encFileName))
        whenever(
            fileResolveHelper.fileExistsAndHasExtension(
                eq(AppProperties.ENC_FILE_EXTENSION),
                any()
            )
        ).thenReturn(true)
        whenever(settings.getValue(eq(SettingsDao.DEC_DIR))).thenReturn("decDir")

        decrypt.execute(args)

        val fileCaptor = argumentCaptor<File>()
        verify(fileResolveHelper).fileExistsAndHasExtension(
            eq(AppProperties.ENC_FILE_EXTENSION),
            any()
        )
        verify(cryptoHelper).decryptFile(fileCaptor.capture(), eq("decDir"), eq(silentFlag))
        assertEquals(encFileName, fileCaptor.firstValue.name)
    }
}